¡Hola, Andrea!

Mi nombre es Tonatiuh Cruz. Me complace revisar tu proyecto hoy.

Al identificar cualquier error inicialmente, simplemente los destacaré. Te animo a localizar y abordar los problemas de forma independiente como parte de tu preparación para un rol como data-scientist. En un entorno profesional, tu líder de equipo seguiría un enfoque similar. Si encuentras la tarea desafiante, proporcionaré una pista más específica en la próxima iteración.

Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o elimines**.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Éxito. Todo está hecho correctamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con comentarios en rojo.
</div>

Puedes responderme utilizando esto:

<div class="alert alert-block alert-info">


# Proyecto SQL

Se analiza la información de uno de los servicios de startup de libros creados durante la pandemia con la finalidad de generar una propuesta de valor para un nuevo producto.

In [1]:
# Se importan las librerías
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

# Código para conectarse a la base de datos

engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [2]:
# Obtención de datos

query_books = '''
SELECT*
FROM books
'''

books = pd.io.sql.read_sql(query_books, con = engine)

print(f'Dataset: Books\n')
display(books.head(5))


Dataset: Books



,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


In [3]:
# Obtención de datos

query_authors = '''
SELECT*
FROM authors
'''

authors = pd.io.sql.read_sql(query_authors, con = engine)

print(f'Dataset: Authors\n')
display(authors.head(5))


Dataset: Authors



,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


In [4]:
# Obtención de datos

query_publishers = '''
SELECT*
FROM publishers
'''

publishers = pd.io.sql.read_sql(query_publishers, con = engine)

print(f'Dataset: Publishers\n')
display(publishers.head(5))


Dataset: Publishers



,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


In [5]:
# Obtención de datos

query_ratings = '''
SELECT*
FROM ratings
'''

ratings = pd.io.sql.read_sql(query_ratings, con = engine)

print(f'\nDataset: Ratings\n')
display(ratings.head(5))



Dataset: Ratings



,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


In [6]:
# Obtención de datos

query_reviews = '''
SELECT*
FROM reviews
'''

reviews = pd.io.sql.read_sql(query_reviews, con = engine)

print(f'Dataset: Reviews\n')
display(reviews.head(5))


Dataset: Reviews



,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


### Encontrar número de libros publicados después del 1 de enero de 2000

In [7]:
# Se cuenta la cantidad de libros publicados después del 1 de enero de 2000

query_books_published = '''
SELECT COUNT(*)
FROM books
WHERE publication_date > '2000-01-01'
'''

books_count = pd.io.sql.read_sql(query_books_published, con = engine)
books_total = books_count.iloc[0,0]

print(f'La cantidad de libros publicados después del 1 de enero de 2000 es de:\n{books_total} libros')


La cantidad de libros publicados después del 1 de enero de 2000 es de:
819 libros


### Comentarios:

La cantidad de libros publicados despues del 1 de enero de 2000 fue de 819 libros. 

Este dato ayuda a formarse una idea de mas o menos cuantos libros pueden estarse publciando anualmente y así poder formar una estrategia para el nuevo lanzamiento.

### Encontrar el número de reseñas de usuarios y la calificación promedio para cada libro

In [8]:
# Se junta la tabla 'books' con 'ratings' y con 'reviews' para poder hacer la consulta
# Se cuenta el número de reseñas de los usurios y la califiación promedio para cada libro

query_reviews_ratings = '''
SELECT
    b.book_id,
    b.title,
    COUNT(rv.review_id) AS num_reviews,
    AVG(rt.rating) AS avg_rating
FROM
    books b
LEFT JOIN
    reviews rv ON b.book_id = rv.book_id
LEFT JOIN
    ratings rt ON b.book_id = rt.book_id
GROUP BY
    b.book_id, b.title
ORDER BY
    avg_rating DESC
'''

books_reviews_ratings = pd.io.sql.read_sql(query_reviews_ratings, con = engine)

print(f'Tabla con la cantidad de reseñas por libro [num_reviews] y calificación\npromedio [avg_rating]\n')
display(books_reviews_ratings)


Tabla con la cantidad de reseñas por libro [num_reviews] y calificación
promedio [avg_rating]



,book_id,title,num_reviews,avg_rating
0,518,Pop Goes the Weasel (Alex Cross #5),4,5.00
1,732,The Ghost Map: The Story of London's Most Terr...,4,5.00
2,347,In the Hand of the Goddess (Song of the Liones...,6,5.00
3,610,Tai-Pan (Asian Saga #2),4,5.00
4,330,How to Be a Domestic Goddess: Baking and the A...,2,5.00
...,...,...,...,...
995,915,The World Is Flat: A Brief History of the Twen...,12,2.25
996,371,Junky,4,2.00
997,316,His Excellency: George Washington,4,2.00
998,202,Drowning Ruth,9,2.00


### Comentarios:

Con esta consulta que está acomodada de forma descendente según el promedio de calificación que tiene cada libro se pueden observar algunas cosas.

Por ejemplo, los libros que tienen una calificaciín más alta no tiene un gran número de reseñas, esto puede indicar que podría haber suarios que califican el titulo sin dejar ningún comentario al respecto o que no muchos conocen el libro/autor, sin embargio a quienes lo conocen y lo han leido les ha encantado y por eso puntuan con la mayor calificación.

Aquellos libros que tienen muchas reseñas y una calificación promedio baja probablemente indican que son libros populares pero no han logrado satisfacer las expectaitvas que tiene el público en generla respecto a ellos.

### Identificar editorial que ha publicado mayor número de libros con más de 50 páginas

In [9]:
# Se juntan las tablas de 'books' y 'publishers' para poder hacer la consulta
# Se identifican las editoriales que han publicado el mayor número de libros con más de 50 páginas

query_books_publisher = '''
SELECT
    p.publisher,
    COUNT(b.title) AS publisher_count
FROM
    publishers p
JOIN
    books b ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY
    p.publisher
ORDER BY
    publisher_count DESC
'''

books_publishers = pd.io.sql.read_sql(query_books_publisher, con = engine)
publisher_name = books_publishers.iloc[0,0]
publisher_count = books_publishers.iloc[0,1]

print(f'La editorial que ha publicado mayor número de libros con más de 50 páginas es: {publisher_name} con {publisher_count} libros publicados')
print(f'\nTabla con datos:\n')
display(books_publishers)


La editorial que ha publicado mayor número de libros con más de 50 páginas es: Penguin Books con 42 libros publicados

Tabla con datos:



,publisher,publisher_count
0,Penguin Books,42
1,Vintage,31
2,Grand Central Publishing,25
3,Penguin Classics,24
4,Ballantine Books,19
...,...,...
329,Turtleback,1
330,Atheneum Books for Young Readers: Richard Jack...,1
331,Penguin Signet,1
332,Victor Gollancz,1


### Comentarios:

La editorial 'Penguin Books' es quien tiene una mayor cantidad de libros publicados con más de 50 páginas.

Para quienes son ávidos lectores, se sabe que le editorial 'Penguin' es una de las más grandes de todas, por lo que no es de sorprender que se encuentre en el lugar en el que está, incluso se puede apreciar en la previsualización de la tabla que hay otra categoria llamada 'Penguin Classics' la cual, aunque esta separa en el nombre se sabe que es también parte de la familia 'Penguin' por lo que hace que el número se eleve todavía más si es que en algún momento del análsis se tuvieran que sumar.

Las diferencias que se pueden aprecair entre las primeras editoriales y las últimas puede deberse a que muy probablemente, los titulos correspondientes a las últimas editoriales sean libros no muy conocidos o textos que no son 'novelas', tal ves alguna enciclopedia o diccionario, lo que hace que las editoriales puedan llegar a no tener un gran interes en su publicación.

# Identificar al autor que tiene la más alta calificación promedio del libro

Se consideran sólo aquellos libros que tienen al menos 50 reseñas.

In [10]:
# Se juntan las tablas 'books', 'ratings' y 'authors' para la consulta
# Se identifica al autor que tiene la más alta calificación promedio del libro
# y se toma en cuenta sólo aquellos libros que tienen al menos 50 reseñas

query_author_rating = '''
SELECT
    a.author,
    COUNT(rt.rating) AS rating_count,
    AVG(rt.rating) AS rating_avg
FROM
    books b
JOIN
    ratings rt ON rt.book_id = b.book_id
JOIN
    authors a ON a.author_id = b.author_id
GROUP BY a.author
HAVING COUNT(rt.rating) >= 50
ORDER BY rating_avg DESC
'''

author_rating = pd.io.sql.read_sql(query_author_rating, con = engine)
author_name = author_rating.iloc[0,0]
author_rating_avg = author_rating.iloc[0,2]

print(f'El autor que tiene la más alta calificación promedio del libro es:\n{author_name} con una califciación promedio de {author_rating_avg}')
print(f'\nTabla con datos:\n')
display(author_rating)


El autor que tiene la más alta calificación promedio del libro es:
Diana Gabaldon con una califciación promedio de 4.3

Tabla con datos:



,author,rating_count,rating_avg
0,Diana Gabaldon,50,4.300000
1,J.K. Rowling/Mary GrandPré,312,4.288462
2,Agatha Christie,53,4.283019
3,Markus Zusak/Cao Xuân Việt Khương,53,4.264151
4,J.R.R. Tolkien,166,4.240964
5,Roald Dahl/Quentin Blake,62,4.209677
6,Louisa May Alcott,54,4.203704
7,Rick Riordan,84,4.130952
8,Arthur Golden,56,4.107143
9,Stephen King,106,4.009434


### Comentarios:

Con la consulta anterior se puede identificar de manera muy sencilla, cual es el autor que tiene un mejor promedio en calificación respecto a sus libros con más de 50 reseñas en la aplicación.

Se observa que no siempre los autores con mayor cantidad de reseñas son quienes tienen una puntuación mucho mayor. Con frecuencia los usuarios tienden a dejar más malos comentarios que buenos, por lo tanto puede ser que una gran cantidad de reseñas no siempre signifique que un libro es bueno o muy popular.

Muchos autores y libros tienen una base fe fans que son muy leales y por ende califican alto los titulos, por ejemplo el caso de J.K. Rowling con 'Harry Potter', quien a pesar de ser una autora con tanto éxito y que sus libros han marcado a toda una generación no tiene la puntuación más alta a pesar de encabezar el número de reseñas.

Además es importante tomar en cuenta, como ya se mencionó con anterioridad que muchas veces los usuarios pueden calificar el titulo, ya se de forma positiva o negativa, pero no dejan reseñas o coemntarios del mismom, por lo que guiarse sólo en la cantidad de comentarios para tomar cualquier decisión no es lo más recomendable.

In [11]:
# Se juntan las tablas de 'ratings' y 'review' para hacer la consulta
# Se encuentra el número promedio de reseñas de texto entre usuarios que calificaron más de 50 libros


query_ratings_review = '''
SELECT
    AVG(total_reviews)
FROM(
SELECT
    rt.username,
    rv.total_reviews
FROM (
    SELECT
        username,
    COUNT(*)
        AS total_ratings
    FROM
        ratings
    GROUP BY
        username
    HAVING
        COUNT(*) > 50
) rt
INNER JOIN (
    SELECT
        username,
    COUNT(*)
        AS total_reviews
    FROM
        reviews
    GROUP BY
        username
) rv ON rt.username = rv.username
) AS subquery
'''


rating_review_avg = pd.io.sql.read_sql(query_ratings_review, con = engine)
text_rating_avg = rating_review_avg.iloc[0, 0]

print(f'El promedio de número de reseñas de texto entre los usuarios que calificaron\nmás de 50 libros es de: {text_rating_avg} reseñas')


El promedio de número de reseñas de texto entre los usuarios que calificaron
más de 50 libros es de: 24.333333333333332 reseñas


### Comentarios:

Siguiendo con los comentarios anteriores, aquí se puede observar que la cantidad promedio de reseñas que dejan los usuarios que han calificado más de 50 libros es de 24.

Esto quiere decir que los usuarios sólo dejan una reseña o comentario a 1/4 parte de los libros que leen y que califican.

### Conclusiones:

El ejercicio no especifica exactamente cuál es la propuesta que se tiene para el nuevo producto o qué es lo que especificamente necesitan del análisis, sin embargo si se puede mencionar que para tomar cualquier tipo de decisión no se recomienda basarse unica y exclusivamente en la cantidad de comentarios que los usuarios dejen en los titulos, ya que la gran mayoria no siempre deja reseñas.

Es más recomendable que se basen el las califiaciones promedio obtenidas para saber cuáles son los autores que estpan teniendo una mayor popularidad y según el género que maneje cada uno decidir cuál sería la mejor opción a elegir, ya sea para saber que próximo audiolibro o libro digital lanzar para que sea mejor recibido por los usarios de la aplicación.

<div class="alert alert-block alert-success">
<b>Paso 2</b> <a class="tocSkip"></a><br>
Has realizado un buen trabajo al formular una consulta SQL para contar con:
    
    - el número de libros publicados después de una fecha específica.
    - Calificación promedio
    - Editorial que ha publicado la mayor cantidad de libros
    - Autor con la calificación promedio más alta
    - Promedio de reseñas
    
Esto demuestra tu habilidad para manipular y consultar bases de datos de manera efectiva. Solamente para terminar de complementar te recomiendo incluir el número de libros evaluados por autor para proporcionar un contexto adicional sobre la consistencia de sus calificaciones.
</div>